extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java

```diff
@@ -5,54 +5,316 @@
+  public <T, E extends Throwable> T call(
+      SpanBuilder spanBuilder,
+      SpanCallable<T, E> spanCallable,
+      BiConsumer<Span, Throwable> handleException)
+      throws E {
+    Span span = spanBuilder.startSpan();
+    //noinspection unused
+    try (Scope unused = span.makeCurrent()) {
+      return spanCallable.doInSpan();
     } catch (Throwable e) {
-      span.recordException(e);
+      handleException.accept(span, e);
       throw e;
     } finally {
       span.end();
     }
   }
 
-  /** Call the provided {@link Callable} and wrap with a {@link Span} with the provided name. */
-  public <T> T call(String spanName, Callable<T> callable) throws Exception {
-    Span span = delegate.spanBuilder(spanName).startSpan();
-    try (Scope scope = span.makeCurrent()) {
-      return callable.call();
-    } catch (Throwable e) {
-      span.recordException(e);
-      throw e;
-    } finally {
-      span.end();
+  /**
+   * Marks a span as error.
+   *
+   * @param span the span
+   * @param exception the exception that caused the error
+   */
+  private static void setSpanError(Span span, Throwable exception) {
+    span.setStatus(StatusCode.ERROR);
+    span.recordException(exception);
+  }
+
+  /**
+   * Set baggage items inside the given {@link SpanCallable}.
+   *
+   * @param baggage the baggage items to set
+   * @param spanCallable the {@link SpanCallable} to call
+   * @param <T> the type of the result
+   * @param <E> the type of the exception
+   * @return the result of the {@link SpanCallable}
+   */
+  @SuppressWarnings("NullAway")
+  public static <T, E extends Throwable> T callWithBaggage(
+      Map<String, String> baggage, SpanCallable<T, E> spanCallable) throws E {
+    BaggageBuilder builder = Baggage.current().toBuilder();
+    baggage.forEach(builder::put);
+    Context context = builder.build().storeInContext(Context.current());
+    try (Scope ignore = context.makeCurrent()) {
+      return spanCallable.doInSpan();
     }
   }
 
-  @Override
-  public SpanBuilder spanBuilder(String spanName) {
-    return delegate.spanBuilder(spanName);
+  /**
+   * Run the given {@link SpanCallable} inside a server span.
+   *
+   * <p>The span context will be extracted from the <code>carrier</code>, which you usually get from
+   * HTTP headers of the metadata of a message you're processing.
+   *
+   * <p>If an exception is thrown by the {@link SpanCallable}, the span will be marked as error, and
+   * the exception will be recorded.
+   *
+   * @param carrier the string map where to extract the span context from
+   * @param spanBuilder the {@link SpanBuilder} to use
+   * @param spanCallable the {@link SpanCallable} to call
+   * @param propagators provide the propagators from {@link OpenTelemetry#getPropagators()}
+   * @param <T> the type of the result
+   * @param <E> the type of the exception
+   * @return the result of the {@link SpanCallable}
+   */
+  public <T, E extends Throwable> T traceServerSpan(
```

Comment:
A couple of thoughts I don't feel overly strong about:
- What about a rename to `extractAndCall`? The difference between this functionality and the `call` methods is that context is extracted first, and the name of the method doesn't really imply that.
- What about switching `ContextPropagators` to the first argument? The argument order would be `propagators, carrier, spanBuilder, spanCallable, handleException`, which matches the argument order of the `call` family of methods and would allow for consistent overloads.
- What do you think about having a single method instead of overloads for the different span kinds? Not sure the overloads are that much easier than providing a span builder with the span kind set, and increase the surface area quite a bit.

Reply:
I really like it. Here's what I suggest:

```java
 public static <T, E extends Throwable> T extractAndCall(
      ContextPropagators propagators,
      Map<String, String> carrier,
      SpanBuilder spanBuilder,
      SpanKind spanKind,
      SpanCallable<T, E> spanCallable)
      throws E {
    return extractAndCall(propagators, carrier, spanBuilder, spanKind, spanCallable,  ExtendedTracer::setSpanError);
  }
```

- guides the user to remember span kind
- there's still an overload `handleException`, but that's only an extra arg

If we want to make it even better, we can use a builder. I think I'll try that out.

Reply:
Here's how it would roughly look like - I actually like it:

```java
                        t.extract(
                            otelTesting.getOpenTelemetry().getPropagators(),
                            Collections.emptyMap()).spanKind(SpanKind.CONSUMER).exceptionHandler(ignoreException).named("span").call(c)
```

Reply:
I've incorporated all other changes, but here I'm waiting for some feedback before implementing the builder pattern.

Reply:
@jack-berg wdyt?

Reply:
The example above is now

`                        t.spanBuilder("span")
                            .setSpanKind(SpanKind.SERVER)
                            .extractAndCall(
                                otelTesting.getOpenTelemetry().getPropagators(),
                                Collections.emptyMap(),
                                c)`